numpy/_array_api/_array_object.py

```diff
@@ -0,0 +1,662 @@
+        """
+        Performs the operation __rmul__.
+        """
+        res = self._array.__rmul__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __ior__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __ior__.
+        """
+        res = self._array.__ior__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __ror__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __ror__.
+        """
+        res = self._array.__ror__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __ipow__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __ipow__.
+        """
+        res = self._array.__ipow__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __rpow__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __rpow__.
+        """
+        res = self._array.__rpow__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __irshift__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __irshift__.
+        """
+        res = self._array.__irshift__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __rrshift__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __rrshift__.
+        """
+        res = self._array.__rrshift__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __isub__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __isub__.
+        """
+        res = self._array.__isub__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __rsub__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __rsub__.
+        """
+        res = self._array.__rsub__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __itruediv__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __itruediv__.
+        """
+        res = self._array.__itruediv__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __rtruediv__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __rtruediv__.
+        """
+        res = self._array.__rtruediv__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __ixor__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __ixor__.
+        """
+        res = self._array.__ixor__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    def __rxor__(self: array, other: array, /) -> array:
+        """
+        Performs the operation __rxor__.
+        """
+        res = self._array.__rxor__(asarray(other)._array)
+        return self.__class__._new(res)
+
+    @property
+    def dtype(self):
```

Comment:
Note to self: this is due to a limitation in how I parsed the spec to generate the types. The return types for the properties are present in the spec (the stubs in the test suite also needs to be updated).

Reply:
> Are there any plans for making this array class generic w.r.t. its shape and dtype?

I don't know what this means (I'm not too knowledgeable about type stuff).

Reply:
This really should be `Literal[int8, int16, ..., float32, float64]`. I think mypy doesn't support this yet, but it's a more correct annotation.

Reply:
Actually this was already addressed here https://github.com/numpy/numpy/pull/18585#discussion_r591762188 The type should be the dtype type from the types file. I also realized that the types should be CamelCased to avoid confusion between the types and the properties.

Reply:
You could indeed return a big union (_e.g._ `np.dtype[int8 | int16 | ... | float32 | float64]`), 
but due to somewhat technical typing-related reasons [1] I'd strongly advice against doing that and 
instead just stick to `np.dtype[Any]`. 

Namelly, you'd be forcing end-users to perform an `isinstance` check on the output of `.dtype`, 
lest they'd be stuck with a weird (and practically useless) superposition of all valid dtypes.

[1] Namelly, _all_ operations performed on a union must be compatible with _all_ of its members,
so you're very quickly gona get into trouble:
``` python
from __future__ import annotations

import numpy as np
from numpy._array_api._array_object import ndarray

def ge_dtype_i8(int_array: ndarray) -> np.dtype[np.int64]:
    # error: Incompatible return value type (got "dtype[Union[signedinteger[_8Bit], signedinteger[_16Bit], signedinteger[_64Bit], floating[_32Bit], floating[_64Bit]]]", 
    # expected "dtype[signedinteger[_64Bit]]")
    return int_array.dtype
```

Reply:
Do you mean this will cause problems for the functions that only accept a subset of dtypes? I can't figure out how to run mypy on the array API submodule.

In terms of operations, the dtype objects in the array API shouldn't have any operations defined on them. The only thing that they need to define is equality checking. It was suggested at some point to use custom wrapper classes for them, but I think @rgommers and @seberg didn't want to do this.